### PR TITLE
feat(iOS): add all supported ReturnKeyTypes

### PR DESCRIPTION
**Pull request by elencho**
## Summary:
Related issue: https://github.com/facebook/react-native/issues/43243
As [documentation](https://reactnative.dev/docs/textinput#returnkeytype) stated for the ReturnKeyType prop on the input there are different options, like "next", "go" and etc. They are actually supported on the iOS side but we can't use it in our react native app, because of the hardcoded version of DoneButton on existing code: 
image
So, I decided to add support for types which were in documentation

## Changelog:
[IOS] [ADDED]: ReturnKeyTypes

## Test Plan:

ran yarn jest react-native-codegen and yarn jest react-native, both successfully:
image

image

### CODE REVIEW

**Review by cipolleschi — APPROVED** [review]
Hi @elencho, thanks for the contribution! I'll import it and see how it behaves internally!

File: packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.mm
Comment (cipolleschi) [inline packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.mm:709]: A couple of suggestions here:
1. Can we create a `NSSet<NSNumber *>` instead of an `NSArray`? This will be more efficient when looking up as we won't have to iterate over the array
2. Can we move this to a static variable out of the implementation?
(before the `@implementation` keyword on top)?

The reason for 2 is that this Collection will be created and destroyed every time the method `setDefaultInputAccessoryView` is invoked, so it is not very efficient.

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Sorry for getting back to this after the approval.
A colleague was doing a second pass and he raised some relevant concerns.

I left a couple of suggestions in the code.

Plus, this is a file for the old architecture only. It is great that you fixed this, but could you apply the same changes to [this other file](https://github.com/facebook/react-native/blob/5a0ae6e2d9d7f2357f9ea6c5dc1d573233075326/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm?fbclid=IwAR28O1oBYdy_bhOYvKzInJ6X6M73f_AWtg_7gMIXwfvIJU5OMD7gL4dsP8U#L479), for the New Architecture?

Thank you so much! 🙏

**Comment by cipolleschi** [comment]
@elencho gentle reminder that there are some changes requested.

**Comment by elencho** [comment]
> @elencho gentle reminder that there are some changes requested.

Sorry I just saw your comment, will fix in few days and get back to you. Thanks for your feedback!!

**Comment by elencho** [comment]
NOTE: I removed Default value, because it was causing a bug: when returnKey was undefined from js it was still showing "Default"
image

**Comment by cipolleschi** [comment]
> NOTE: I removed Default value, because it was causing a bug: when returnKey was undefined from js it was still showing "Default"

Why you think it is a bug? I feel like it is the right behavior to fall back to `Default` if nothing is specified, no?

**Comment by elencho** [comment]
> > NOTE: I removed Default value, because it was causing a bug: when returnKey was undefined from js it was still showing "Default"
> 
> Why you think it is a bug? I feel like it is the right behavior to fall back to `Default` if nothing is specified, no?

Okay i will revert

**Review by cipolleschi — APPROVED** [review]

**Comment by cipolleschi** [comment]
/rebase - this comment automatically rebase on top of main

**Comment by cipolleschi** [comment]
@elencho No, nothing wrong.
I have to run the internal linter (which is different from the OSS one, unfortunately) and ask for a review internally.
I'm still catching up from AppJS and the ReactConf, that's why I have been a little slow, these days.

**Comment by kirillzyusko** [comment]
I have a feeling like it had to be removed https://github.com/facebook/react-native/pull/43362#issuecomment-2101581007

Because right now when you didn't specify `returnTypeKey` (i. e. it's undefined) the toolbar always shows "Default" and this button doesn't do anything (i. e. I press, but nothing happens).

I have a strong feeling that it's a regression 😅 Correct me if I'm wrong 😅

**Comment by elencho** [comment]
Hello @kirillzyusko , thank you for feedback. 
Just tested and you are right, I would love to contribute again here. 
Can I investigate this and come back with fixed solution?

**Comment by kirillzyusko** [comment]
> Can I investigate this and come back with fixed solution?

Yes, please 🙏 

I also would like to get @cipolleschi opinion on that - whether it should be considered as a release blocker or not and what is the best way to resolve it here.

The straightforward approach is to remove `UIReturnKeyDefault` from the `Set`. Or if `undefined` or `null` is coming from JS, then do not substitute the value with `UIReturnKeyDefault` by default and allow that value to be `nil`, but it may affect other places of the framework (just afraid to get `NullPointerException` somewhere in the code 😅).

Anyway, it's just suggestions 🙃

**Comment by cipolleschi** [comment]
Hi There, sorry for the issue!

> I also would like to get @cipolleschi opinion on that - whether it should be considered as a release blocker or not and what is the best way to resolve it here.

We cannot stop the release, which is happening tomorrow, at this point, especially for this that has not been reported by anyone else.
However, we are planning todo 75.1 next week, so this fix can go in there.

**Comment by kirillzyusko** [comment]
> We cannot stop the release, which is happening tomorrow, at this point, especially for this that has not been reported by anyone else.

@cipolleschi got it!

> However, we are planning todo 75.1 next week, so this fix can go in there.

That would be amazing to include this fix in 0.75.1 👍 Do you think that removing `UIReturnKeyDefault` from a set can be included in 0.75.1? Or do you want to have a different fix?

**Comment by elencho** [comment]
@kirillzyusko I will try to include it, but if there wont be another fix I will remove that

**Comment by cipolleschi** [comment]
@elencho just to understand, do you plan to look into this issue?

**Comment by elencho** [comment]
> @elencho just to understand, do you plan to look into this issue?

Yes, I will contribute tomorrow

**Pull request closed without merging**
